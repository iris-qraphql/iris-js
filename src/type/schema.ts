import { uniqBy } from 'ramda';

import type { IrisError } from '../error';
import { inspect, instanceOf } from '../utils/legacy';
import type { ObjMap } from '../utils/ObjMap';
import type { Maybe } from '../utils/type-level';

import type { IrisNamedType, IrisResolverType, IrisType } from './definition';
import { getNamedType, isDataType, isResolverType } from './definition';
import type { GraphQLDirective } from './directives';
import { isDirective, specifiedDirectives } from './directives';

/**
 * Test if the given value is a GraphQL schema.
 */
export function isSchema(schema: unknown): schema is IrisSchema {
  return instanceOf(schema, IrisSchema);
}

export function assertSchema(schema: unknown): IrisSchema {
  if (!isSchema(schema)) {
    throw new Error(`Expected ${inspect(schema)} to be a GraphQL schema.`);
  }
  return schema;
}

export class IrisSchema {
  description: Maybe<string>;

  // Used as a cache for validateSchema().
  __validationErrors: Maybe<ReadonlyArray<IrisError>>;

  private _queryType: Maybe<IrisResolverType>;
  private _mutationType: Maybe<IrisResolverType>;
  private _subscriptionType: Maybe<IrisResolverType>;
  private _directives: ReadonlyArray<GraphQLDirective>;
  private _typeMap: TypeMap;

  constructor(config: Readonly<GraphQLSchemaConfig>) {
    this.__validationErrors = config.assumeValid === true ? [] : undefined;

    this.description = config.description;

    this._queryType = config.query;
    this._mutationType = config.mutation;
    this._subscriptionType = config.subscription;
    this._directives = uniqBy(
      (x) => x.name,
      [...(config.directives ?? []), ...specifiedDirectives],
    );

    // To preserve order of user-provided types, we add first to add them to
    // the set of "collected" types, so `collectReferencedTypes` ignore them.
    const allReferencedTypes: Set<IrisNamedType> = new Set(config.types);

    config.types?.forEach((type) => {
      allReferencedTypes.delete(type);
      collectReferencedTypes(type, allReferencedTypes);
    });

    [this._queryType, this._mutationType, this._subscriptionType].forEach(
      (type) =>
        type ? collectReferencedTypes(type, allReferencedTypes) : undefined,
    );

    for (const directive of this._directives) {
      // Directives are not validated until validateSchema() is called.
      if (isDirective(directive)) {
        for (const arg of directive.args) {
          collectReferencedTypes(arg.type, allReferencedTypes);
        }
      }
    }

    // Storing the resulting map for reference by the schema.
    this._typeMap = Object.create(null);

    for (const namedType of allReferencedTypes) {
      if (namedType == null) {
        continue;
      }

      const typeName = namedType.name;
      if (!typeName) {
        throw new Error(
          'One of the provided types for building the Schema is missing a name.',
        );
      }

      if (this._typeMap[typeName] !== undefined) {
        throw new Error(
          `Iris Schema must contain uniquely named types but contains multiple types named "${typeName}".`,
        );
      }
      this._typeMap[typeName] = namedType;
    }
  }

  get [Symbol.toStringTag]() {
    return 'IrisSchema';
  }

  getQueryType(): IrisResolverType | undefined {
    return this._queryType ?? undefined;
  }

  getMutationType(): IrisResolverType | undefined {
    return this._mutationType ?? undefined;
  }

  getSubscriptionType(): IrisResolverType | undefined {
    return this._subscriptionType ?? undefined;
  }

  getTypeMap = (): TypeMap => this._typeMap;

  getType(name: string): IrisNamedType | undefined {
    return this.getTypeMap()[name];
  }

  getDirectives(): ReadonlyArray<GraphQLDirective> {
    return this._directives;
  }

  getDirective(name: string): Maybe<GraphQLDirective> {
    return this.getDirectives().find((directive) => directive.name === name);
  }
}

type TypeMap = ObjMap<IrisNamedType>;

export type IrisSchemaValidationOptions = {
  assumeValid?: boolean;
  assumeValidSDL?: boolean;
};

export interface GraphQLSchemaConfig extends IrisSchemaValidationOptions {
  description?: Maybe<string>;
  query?: Maybe<IrisResolverType>;
  mutation?: Maybe<IrisResolverType>;
  subscription?: Maybe<IrisResolverType>;
  types?: Maybe<ReadonlyArray<IrisNamedType>>;
  directives?: Maybe<ReadonlyArray<GraphQLDirective>>;
}

function collectReferencedTypes(
  type: IrisType,
  typeSet: Set<IrisNamedType>,
): Set<IrisNamedType> {
  const namedType = getNamedType(type);

  if (!typeSet.has(namedType)) {
    typeSet.add(namedType);
    if (isResolverType(namedType)) {
      const variants = namedType.variants();
      if (namedType.isVariantType()) {
        const fields = Object.values(variants[0]?.fields ?? {});
        for (const field of fields) {
          collectReferencedTypes(field.type, typeSet);
          for (const arg of field.args) {
            collectReferencedTypes(arg.type, typeSet);
          }
        }
      } else {
        for (const variant of variants) {
          if (variant.type) {
            collectReferencedTypes(variant.type, typeSet);
          }
        }
      }
    } else if (isDataType(namedType)) {
      namedType
        .variants()
        .flatMap((x) => Object.values(x.fields ?? {}))
        .forEach((field) => collectReferencedTypes(field.type, typeSet));
    }
  }

  return typeSet;
}
