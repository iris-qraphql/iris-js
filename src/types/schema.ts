import type { DirectiveNode, ParseOptions, Source } from 'graphql';
import { prop, uniqBy } from 'ramda';

import { typeCheckASTValue } from '../validation/typeCheckASTValue';
import { validateSDL } from '../validation/validate';

import type { IrisError } from '../error';
import { irisNodeError } from '../error';
import { parse } from '../parsing';
import type { TypeMap } from '../utils/collectTypeMap';
import { collectTypeMap } from '../utils/collectTypeMap';
import { inspect } from '../utils/legacy';
import type { IrisMaybe, Maybe } from '../utils/type-level';
import { notNill } from '../utils/type-level';

import type {
  ArgumentDefinitionNode,
  DirectiveDefinitionNode,
  FieldDefinitionNode,
  NamedTypeNode,
  Role,
  TypeDefinitionNode,
  TypeNode,
  VariantDefinitionNode,
} from './ast';
import type {
  IrisArgument,
  IrisField,
  IrisType,
  IrisVariant,
} from './definition';
import { IrisScalars, IrisTypeDefinition, IrisTypeRef } from './definition';
import {
  GraphQLDeprecatedDirective,
  GraphQLDirective,
  specifiedDirectives,
} from './directives';
import { IrisKind } from './kinds';

class IrisSchema {
  description: Maybe<string>;

  readonly query?: IrisTypeDefinition<'resolver'>;
  readonly mutation?: IrisTypeDefinition<'resolver'>;
  readonly subscription?: IrisTypeDefinition<'resolver'>;
  readonly directives: ReadonlyArray<GraphQLDirective>;
  readonly typeMap: TypeMap;

  // Used as a cache for validateSchema().
  __validationErrors: Maybe<ReadonlyArray<IrisError>>;

  constructor({
    description,
    query,
    mutation,
    subscription,
    types = [],
    directives = [],
  }: Readonly<IrisSchemaConfig>) {
    this.description = description;
    this.query = query;
    this.mutation = mutation;
    this.subscription = subscription;
    this.directives = uniqBy(prop('name'), [
      ...directives,
      ...specifiedDirectives,
    ]);

    this.typeMap = collectTypeMap(
      [query, mutation, subscription, ...types].filter(notNill),
      this.directives,
    );
  }

  get [Symbol.toStringTag]() {
    return 'IrisSchema';
  }

  getType = (name: string): IrisMaybe<IrisTypeDefinition> => this.typeMap[name];

  getDirective = (name: string): Maybe<GraphQLDirective> =>
    this.directives.find((directive) => directive.name === name);
}

export interface IrisSchemaConfig {
  description?: Maybe<string>;
  query?: IrisTypeDefinition<'resolver'>;
  mutation?: IrisTypeDefinition<'resolver'>;
  subscription?: IrisTypeDefinition<'resolver'>;
  types?: ReadonlyArray<IrisTypeDefinition>;
  directives?: ReadonlyArray<GraphQLDirective>;
}

export function buildSchema(
  source: string | Source,
  options?: ParseOptions,
): IrisSchema {
  const documentAST = parse(source, { noLocation: options?.noLocation });
  const errors = validateSDL(documentAST);

  if (errors.length !== 0) {
    throw new Error(errors.map((error) => error.message).join('\n\n'));
  }

  const directiveDefs: Array<DirectiveDefinitionNode> = [];
  const typeMap: Record<string, IrisTypeDefinition> = {};

  function lookupType<R extends Role>(
    node: NamedTypeNode | VariantDefinitionNode<R>,
  ): IrisTypeDefinition<R> {
    const name = node.name.value;
    const type = IrisScalars[name] ?? typeMap[name];

    if (type === undefined) {
      throw new Error(`Unknown type: "${name}".`);
    }

    return type as IrisTypeDefinition<R>;
  }

  function getWrappedType(node: TypeNode): IrisType {
    if (node.kind === IrisKind.LIST_TYPE) {
      return new IrisTypeRef('LIST', getWrappedType(node.type));
    }
    if (node.kind === IrisKind.MAYBE_TYPE) {
      return new IrisTypeRef('MAYBE', getWrappedType(node.type));
    }
    return lookupType(node);
  }

  function buildDirective(node: DirectiveDefinitionNode): GraphQLDirective {
    return new GraphQLDirective({
      name: node.name.value,
      description: node.description?.value,
      locations: node.locations.map(({ value }) => value as any),
      isRepeatable: node.repeatable,
      args: node.arguments?.map(buildArgument),
      astNode: node,
    });
  }

  function buildArgument(astNode: ArgumentDefinitionNode): IrisArgument {
    const type: any = getWrappedType(astNode.type);
    const name = astNode.name.value;
    return {
      name,
      type,
      description: astNode.description?.value,
      defaultValue: typeCheckASTValue(astNode.defaultValue, type),
      deprecationReason: getDeprecationReason(astNode),
      astNode,
    };
  }

  const buildField = <R extends Role>(
    field: FieldDefinitionNode<R>,
  ): IrisField<R> =>
    ({
      name: field.name.value,
      type: getWrappedType(field.type),
      description: field.description?.value,
      deprecationReason: getDeprecationReason(field),
      astNode: field,
      args: field.arguments?.map(buildArgument),
      toJSON: () => field.name.value,
    } as IrisField<R>);

  const buildVariant = <R extends Role>(
    astNode: VariantDefinitionNode<R>,
  ): IrisVariant<R> => {
    const name = astNode.name.value;
    const description = astNode.description?.value;
    const deprecationReason = getDeprecationReason(astNode);

    if (!astNode.fields) {
      return {
        name,
        description,
        deprecationReason,
        astNode,
        type: lookupType(astNode),
      };
    }

    return {
      name,
      description,
      deprecationReason,
      fields: Object.fromEntries(
        (astNode.fields ?? []).map((field) => [
          field.name.value,
          buildField(field),
        ]),
      ),
      astNode,
    };
  };

  const buildType = <R extends Role>(
    astNode: TypeDefinitionNode<R>,
  ): IrisTypeDefinition<R> =>
    new IrisTypeDefinition({
      role: astNode.role,
      name: astNode.name.value,
      description: astNode.description?.value,
      variants: () => astNode.variants.map(buildVariant),
      astNode,
    });

  documentAST.definitions.forEach((def) => {
    const name = def.name.value;
    switch (def.kind) {
      case IrisKind.TYPE_DEFINITION:
        typeMap[name] = IrisScalars[name] ?? buildType(def);
        break;
      case IrisKind.DIRECTIVE_DEFINITION:
        directiveDefs.push(def);
        break;
    }
  });

  const assertRoot = (
    operation: string,
    type: IrisTypeDefinition,
  ): IrisTypeDefinition<'resolver'> => {
    if (type && !(type.role === 'resolver' && type.isVariantType)) {
      throw irisNodeError(
        `${operation} root type must be Object type${
          operation === 'Query' ? '' : ' if provided'
        }, it cannot be ${inspect(type)}.`,
        type.astNode,
      );
    }
    return type as IrisTypeDefinition<'resolver'>;
  };

  return new IrisSchema({
    description: undefined,
    types: Object.values(typeMap),
    query: assertRoot('query', typeMap.Query),
    mutation: assertRoot('mutation', typeMap.Mutation),
    subscription: assertRoot('subscription', typeMap.Subscription),
    directives: directiveDefs.map(buildDirective),
  });
}

function getDeprecationReason(node: {
  readonly directives?: ReadonlyArray<DirectiveNode>;
}): IrisMaybe<string> {
  const directiveNode = node.directives?.find(
    (directive) => directive.name.value === GraphQLDeprecatedDirective.name,
  );

  if (directiveNode === undefined) {
    return undefined;
  }

  const reason = directiveNode?.arguments?.find(
    (arg) => arg.name.value === 'reason',
  )?.value;

  return (typeCheckASTValue(reason, IrisScalars.String) as string) ?? '';
}

export type { IrisSchema };
