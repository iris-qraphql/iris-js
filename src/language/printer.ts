import type { Maybe } from '../jsutils/Maybe';

import type { ASTNode } from './ast';
import { printBlockString } from './blockString';
import { printString } from './printString';
import type { ASTReducer } from './visitor';
import { visit } from './visitor';

/**
 * Converts an AST into a string, using one set of reasonable
 * formatting rules.
 */
export function print(ast: ASTNode): string {
  return visit(ast, printDocASTReducer);
}

const MAX_LINE_LENGTH = 80;

const printDocASTReducer: ASTReducer<string> = {
  Name: { leave: (node) => node.value },
  Variable: { leave: (node) => '$' + node.name },

  // Document

  Document: {
    leave: (node) => join(node.definitions, '\n\n'),
  },

  OperationDefinition: {
    leave(node) {
      const varDefs = wrap('(', join(node.variableDefinitions, ', '), ')');
      const prefix = join(
        [
          node.operation,
          join([node.name, varDefs]),
          join(node.directives, ' '),
        ],
        ' ',
      );

      // Anonymous queries with no directives or variable definitions can use
      // the query short form.
      return (prefix === 'query' ? '' : prefix + ' ') + node.selectionSet;
    },
  },

  VariableDefinition: {
    leave: ({ variable, type, defaultValue, directives }) =>
      variable +
      ': ' +
      type +
      wrap(' = ', defaultValue) +
      wrap(' ', join(directives, ' ')),
  },
  SelectionSet: { leave: ({ selections }) => block(selections) },

  Field: {
    leave({ alias, name, arguments: args, directives, selectionSet }) {
      const prefix = wrap('', alias, ': ') + name;
      let argsLine = prefix + wrap('(', join(args, ', '), ')');

      if (argsLine.length > MAX_LINE_LENGTH) {
        argsLine = prefix + wrap('(\n', indent(join(args, '\n')), '\n)');
      }

      return join([argsLine, join(directives, ' '), selectionSet], ' ');
    },
  },

  Argument: { leave: ({ name, value }) => name + ': ' + value },

  // Fragments

  FragmentSpread: {
    leave: ({ name, directives }) =>
      '...' + name + wrap(' ', join(directives, ' ')),
  },

  InlineFragment: {
    leave: ({ typeCondition, directives, selectionSet }) =>
      join(
        [
          '...',
          wrap('on ', typeCondition),
          join(directives, ' '),
          selectionSet,
        ],
        ' ',
      ),
  },

  FragmentDefinition: {
    leave: ({
      name,
      typeCondition,
      variableDefinitions,
      directives,
      selectionSet,
    }) =>
      // Note: fragment variable definitions are experimental and may be changed
      // or removed in the future.
      `fragment ${name}${wrap('(', join(variableDefinitions, ', '), ')')} ` +
      `on ${typeCondition} ${wrap('', join(directives, ' '), ' ')}` +
      selectionSet,
  },

  // Value

  IntValue: { leave: ({ value }) => value },
  FloatValue: { leave: ({ value }) => value },
  StringValue: {
    leave: ({ value, block: isBlockString }) =>
      isBlockString ? printBlockString(value) : printString(value),
  },
  BooleanValue: { leave: ({ value }) => (value ? 'true' : 'false') },
  NullValue: { leave: () => 'null' },
  EnumValue: {
    leave: ({ value }) => value,
  },
  ListValue: { leave: ({ values }) => '[' + join(values, ', ') + ']' },
  ObjectValue: { leave: ({ fields }) => '{' + join(fields, ', ') + '}' },
  ObjectField: { leave: ({ name, value }) => name + ': ' + value },

  // Directive

  Directive: {
    leave: ({ name, arguments: args }) =>
      '@' + name + wrap('(', join(args, ', '), ')'),
  },

  // Type

  NamedType: { leave: ({ name }) => name },
  ListType: { leave: ({ type }) => '[' + type + ']' },
  NonNullType: { leave: ({ type }) => type + '!' },

  // Type System Definitions

  SchemaDefinition: {
    leave: ({ description, directives, operationTypes }) =>
      wrap('', description, '\n') +
      join(['schema', join(directives, ' '), block(operationTypes)], ' '),
  },

  OperationTypeDefinition: {
    leave: ({ operation, type }) => operation + ': ' + type,
  },

  FieldDefinition: {
    leave: ({ description, name, arguments: args, type, directives }) =>
      wrap('', description, '\n') +
      name +
      (hasMultilineItems(args)
        ? wrap('(\n', indent(join(args, '\n')), '\n)')
        : wrap('(', join(args, ', '), ')')) +
      ': ' +
      type +
      wrap(' ', join(directives, ' ')),
  },

  InputValueDefinition: {
    leave: ({ description, name, type, defaultValue, directives }) =>
      wrap('', description, '\n') +
      join(
        [name + ': ' + type, wrap('= ', defaultValue), join(directives, ' ')],
        ' ',
      ),
  },

  ResolverTypeDefinition: {
    leave: ({ description, name, directives, variants }) =>
      wrap('', description, '\n') +
      join(
        [
          'resolver',
          name,
          join(directives, ' '),
          wrap('= ', join(variants, ' | ')),
        ],
        ' ',
      ),
  },

  // ObjectTypeDefinition: {
  //   leave: ({ description, name, directives, fields }) =>
  //     wrap('', description, '\n') +
  //     join(['type', name, join(directives, ' '), block(fields)], ' '),
  // },

  // TODO: consider variants with fields
  VariantDefinition: { leave: ({ name }) => name },

  DataTypeDefinition: {
    leave: ({ description, name, directives, variants }) =>
      wrap('', description, '\n') +
      join(['data', name, join(directives, ' '), block(variants)], ' '),
  },

  DirectiveDefinition: {
    leave: ({ description, name, arguments: args, repeatable, locations }) =>
      wrap('', description, '\n') +
      'directive @' +
      name +
      (hasMultilineItems(args)
        ? wrap('(\n', indent(join(args, '\n')), '\n)')
        : wrap('(', join(args, ', '), ')')) +
      (repeatable ? ' repeatable' : '') +
      ' on ' +
      join(locations, ' | '),
  },
};

/**
 * Given maybeArray, print an empty string if it is null or empty, otherwise
 * print all items together separated by separator if provided
 */
function join(
  maybeArray: Maybe<ReadonlyArray<string | undefined>>,
  separator = '',
): string {
  return maybeArray?.filter((x) => x).join(separator) ?? '';
}

/**
 * Given array, print each item on its own line, wrapped in an indented `{ }` block.
 */
function block(array: Maybe<ReadonlyArray<string | undefined>>): string {
  return wrap('{\n', indent(join(array, '\n')), '\n}');
}

/**
 * If maybeString is not null or empty, then wrap with start and end, otherwise print an empty string.
 */
function wrap(
  start: string,
  maybeString: Maybe<string>,
  end: string = '',
): string {
  return maybeString != null && maybeString !== ''
    ? start + maybeString + end
    : '';
}

function indent(str: string): string {
  return wrap('  ', str.replace(/\n/g, '\n  '));
}

function hasMultilineItems(maybeArray: Maybe<ReadonlyArray<string>>): boolean {
  /* c8 ignore next */
  return maybeArray?.some((str) => str.includes('\n')) ?? false;
}
