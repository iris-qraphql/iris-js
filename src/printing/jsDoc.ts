import { isEmpty } from 'ramda';

import type {
  IrisField,
  IrisTypeDefinition,
  IrisTypeRef,
  IrisVariant,
} from '../types/definition';

const variantTypeName = (typeName: string, name: string) =>
  `${typeName}_${name}`;

const renderVariant =
  (typeName: string) =>
  ({ name, type, fields }: IrisVariant<'data'>) => {
    if (type) {
      return name;
    }

    if (isEmpty(fields)) {
      return `"${name}"`;
    }

    return variantTypeName(typeName, name);
  };

const isStandaloneVariantType = ({ type, fields }: IrisVariant<'data'>) =>
  !isEmpty(fields) && !type;

const renderTypeRef = (type: IrisTypeRef<'data'>): string => {
  switch (type.kind) {
    case 'MAYBE':
      return `?${renderTypeRef(type.ofType)}`;

    case 'LIST':
      return `${renderTypeRef(type.ofType)}[]`;
    case 'NAMED':
      return type.ofType.name;
  }
};

const renderField = (name: string, { type }: IrisField<'data'>) =>
  `${name}: ${renderTypeRef(type)}`;

const renderVariantType =
  (typeName: string) =>
  ({ name, fields }: IrisVariant<'data'>) => {
    const fieldEntries = Object.entries(fields ?? {});
    const defName = variantTypeName(typeName, name);

    const fieldDefs = [
      `__typename: "${name}"`,
      ...fieldEntries.map((arg) => renderField(...arg)),
    ];

    return `@typedef {{ ${fieldDefs.join(', ')} }} ${defName}`;
  };

export const toJSODoc = (type: IrisTypeDefinition<'data'>) => {
  const variants = type.variants();
  const variantTypes = variants
    .filter(isStandaloneVariantType)
    .map(renderVariantType(type.name));

  const variantRefs = variants.map(renderVariant(type.name)).join(' | ');
  return [...variantTypes, `@type {(${variantRefs})}`].join('\n');
};
