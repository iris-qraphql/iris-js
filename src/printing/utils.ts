import type { Maybe } from '../utils/type-level';

/**
 * Given maybeArray, print an empty string if it is null or empty, otherwise
 * print all items together separated by separator if provided
 */
export function join(
  maybeArray: Maybe<ReadonlyArray<string | undefined>>,
  separator = '',
): string {
  return maybeArray?.filter((x) => x).join(separator) ?? '';
}

/**
 * Given array, print each item on its own line, wrapped in an indented `{ }` block.
 */
export const block = (array: ReadonlyArray<string | undefined>) =>
  array.length > 0 ? `{\n${indent(join(array, '\n'))}\n}` : '{}';

/**
 * If maybeString is not null or empty, then wrap with start and end, otherwise print an empty string.
 */
export function wrap(
  start: string,
  maybeString: Maybe<string>,
  end: string = '',
): string {
  return maybeString != null && maybeString !== ''
    ? start + maybeString + end
    : '';
}

export function indent(str: string): string {
  return wrap('  ', str.replace(/\n/g, '\n  '));
}
