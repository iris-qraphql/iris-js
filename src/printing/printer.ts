import { printBlockString } from 'graphql/language/blockString';
import { printString } from 'graphql/language/printString';

import type { ASTNode } from '../types/ast';
import type { ASTReducer } from '../types/visitor';
import { visit } from '../types/visitor';
import type { Maybe } from '../utils/type-level';

/**
 * Converts an AST into a string, using one set of reasonable
 * formatting rules.
 */

const printDocASTReducer: ASTReducer<string> = {
  Name: { leave: (node) => node.value },
  Variable: { leave: (node) => '$' + node.name },
  Document: {
    leave: (node) => join(node.definitions, '\n\n'),
  },
  Argument: { leave: ({ name, value }) => name + ': ' + value },
  IntValue: { leave: ({ value }) => value },
  FloatValue: { leave: ({ value }) => value },
  StringValue: {
    leave: ({ value, block: isBlockString }) =>
      isBlockString ? printBlockString(value) : printString(value),
  },
  BooleanValue: { leave: ({ value }) => (value ? 'true' : 'false') },
  NullValue: { leave: () => 'null' },
  EnumValue: {
    leave: ({ value }) => value,
  },
  ListValue: { leave: ({ values }) => '[' + join(values, ', ') + ']' },
  ObjectValue: { leave: ({ fields }) => '{' + join(fields, ', ') + '}' },
  ObjectField: { leave: ({ name, value }) => name + ': ' + value },

  // Directive

  Directive: {
    leave: ({ name, arguments: args }) =>
      '@' + name + wrap('(', join(args, ', '), ')'),
  },

  // Type

  NamedType: { leave: ({ name }) => name },
  ListType: { leave: ({ type }) => '[' + type + ']' },
  MaybeType: { leave: ({ type }) => type + '?' },

  // Type System Definitions

  FieldDefinition: {
    leave: ({ description, name, type, directives }) =>
      wrap('', description, '\n') +
      name +
      ': ' +
      type +
      wrap(' ', join(directives, ' ')),
  },

  InputValueDefinition: {
    leave: ({ description, name, type, defaultValue, directives }) =>
      wrap('', description, '\n') +
      join(
        [name + ': ' + type, wrap('= ', defaultValue), join(directives, ' ')],
        ' ',
      ),
  },

  TypeDefinition: {
    leave: ({ description, name, directives, variants }) =>
      wrap('', description, '\n') +
      join(
        [
          'data',
          name,
          join(directives, ' '),
          wrap('= ', join(variants, ' | ')),
        ],
        ' ',
      ),
  },

  VariantDefinition: { leave: ({ name }) => name + block([]) },

  DirectiveDefinition: {
    leave: ({ description, name, arguments: args, repeatable, locations }) =>
      wrap('', description, '\n') +
      'directive @' +
      name +
      (hasMultilineItems(args)
        ? wrap('(\n', indent(join(args, '\n')), '\n)')
        : wrap('(', join(args, ', '), ')')) +
      (repeatable ? ' repeatable' : '') +
      ' on ' +
      join(locations, ' | '),
  },
};

/**
 * Given maybeArray, print an empty string if it is null or empty, otherwise
 * print all items together separated by separator if provided
 */
function join(
  maybeArray: Maybe<ReadonlyArray<string | undefined>>,
  separator = '',
): string {
  return maybeArray?.filter((x) => x).join(separator) ?? '';
}

/**
 * Given array, print each item on its own line, wrapped in an indented `{ }` block.
 */
function block(array: Maybe<ReadonlyArray<string | undefined>>): string {
  return wrap('{\n', indent(join(array, '\n')), '\n}');
}

/**
 * If maybeString is not null or empty, then wrap with start and end, otherwise print an empty string.
 */
function wrap(
  start: string,
  maybeString: Maybe<string>,
  end: string = '',
): string {
  return maybeString != null && maybeString !== ''
    ? start + maybeString + end
    : '';
}

function indent(str: string): string {
  return wrap('  ', str.replace(/\n/g, '\n  '));
}

function hasMultilineItems(maybeArray: Maybe<ReadonlyArray<string>>): boolean {
  /* c8 ignore next */
  return maybeArray?.some((str) => str.includes('\n')) ?? false;
}

export const print = (ast: ASTNode): string => visit(ast, printDocASTReducer);
