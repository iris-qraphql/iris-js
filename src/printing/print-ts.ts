import type { Maybe } from 'graphql/jsutils/Maybe';

import type { IrisSchema } from '../types/schema';
import type { ASTReducer } from '../utils/visitor';
import { visit } from '../utils/visitor';

import { block, indent, join, wrap } from './utils';

const mapping: Record<string, Maybe<string>> = {
  String: 'string',
  Int: 'number',
  Float: 'number',
};

const printTypesASTReducer: ASTReducer<string> = {
  Name: { leave: (node) => node.value },
  Document: { leave: (node) => join(node.definitions, '\n\n') },
  NamedType: { leave: ({ name }) => mapping[name] ?? name },
  ListType: { leave: ({ type }) => '[' + type + ']' },
  MaybeType: { leave: ({ type }) => `Maybe<${type}>` },
  FieldDefinition: {
    leave: ({ description, name, type, directives }) =>
      wrap('', description, '\n') +
      name +
      ': ' +
      type +
      wrap(' ', join(directives, ' ')),
  },
  TypeDefinition: {
    leave: ({ description, name, directives, variants }) =>
      wrap('', description, '\n') +
      join(
        [
          'export type',
          name,
          join(directives, ' '),
          variants.length === 1
            ? `= ${variants[0]};`
            : wrap(
                '=\n',
                indent(variants.map((x) => `| ${x}`).join('\n')),
                ';',
              ),
        ],
        ' ',
      ),
  },
  VariantDefinition: {
    leave: ({ name, fields, isTypeVariantNode }) => {
      if (!fields) {
        return name;
      }

      const rendering = block(
        [`__typename: '${name}'`, ...fields].map((x) => `${x};`),
      );

      return isTypeVariantNode ? rendering : indent(rendering).trimStart();
    },
  },
};

const printFunctionsASTReducer: ASTReducer<string> = {
  Name: { leave: (node) => node.value },
  Document: { leave: (node) => join(node.definitions, '\n\n') + '\n' },
  NamedType: { leave: ({ name }) => `iris${name}` },
  ListType: { leave: ({ type }) => '[' + type + ']' },
  MaybeType: { leave: ({ type }) => `irisMaybe(${type})` },
  FieldDefinition: {
    leave: ({ description, name, type, directives }) =>
      wrap('', description, '\n') +
      name +
      ': ' +
      type +
      wrap(' ', join(directives, ' ')),
  },
  TypeDefinition: {
    leave: ({ description, name, variants }) =>
      wrap('', description, '\n') +
      join([
        `export const iris${name} = oneOf<${name}>({\n${indent(
          variants.map((x) => x + ',').join('\n'),
        )}\n});`,
      ]),
  },
  VariantDefinition: {
    leave: ({ name, fields }) =>
      `${name}: ${
        fields
          ? `irisVariant(${block(fields.map((x) => `${x},`))})`
          : `iris${name}`
      }`,
  },
};

const inlineUtils = `
import type { Maybe } from '../scripts/utils';
import {
  irisFloat,
  irisInt,
  irisMaybe,
  irisString,
  irisVariant,
  oneOf,
} from '../scripts/utils';
`;

export const toTSDefinitions = (ast: IrisSchema): string =>
  inlineUtils.trimStart() +
  '\n' +
  visit(ast.document, printTypesASTReducer) +
  '\n\n' +
  visit(ast.document, printFunctionsASTReducer);
