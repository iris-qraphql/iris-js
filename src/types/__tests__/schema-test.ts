import { print } from '../../printing/printer';
import { dedent } from '../../utils/dedent';

import { buildSchema } from '../schema';

const cycle = (src: string) =>
  expect(print(buildSchema(src))).toEqual(dedent([src]));

describe('Type System: Schema', () => {
  it('Define sample schema', () => {
    cycle(`
      resolver Query = {
        article(id: String): Article
        feed: [Article]
      }

      resolver Article = {
        id: String
        isPublished: Boolean
        author: Author
        title: String
        body: String
      }

      resolver Author = {
        id: String
        name: String
        pic(width: Int, height: Int): Image
        recentArticle: Article
      }

      resolver Image = {
        url: String
        width: Int
        height: Int
      }

      resolver Mutation = {
        writeArticle: Article
      }

      resolver Subscription = {
        articleSubscribe(id: String): Article
      }
    `);
  });

  describe('Type Map', () => {
    it('includes data types only used in directives', () => {
      const schema = buildSchema(`
       data Foo 
       data Bar
       resolver Query = {}
      `);

      expect(Object.keys(schema.types)).toEqual(
        expect.arrayContaining(['Foo', 'Bar']),
      );
    });
  });

  describe('Validity', () => {
    describe('A Schema must contain uniquely named types', () => {
      // TODO:
      // it('rejects a Schema which redefines a built-in type', () => {
      //   const schema = buildSchema(`
      //       data String
      //       resolver Query = {
      //         fakeString: String
      //       }
      //   `);

      //   expect(schema).toEqual(
      //     'Schema must contain uniquely named types but contains multiple types named "String".',
      //   );
      // });

      it('rejects a Schema which defines an object type twice', () => {
        expect(() =>
          buildSchema(`
            resolver SameName
            resolver SameName
          `),
        ).toThrowErrorMatchingSnapshot();
      });
    });
  });
});
