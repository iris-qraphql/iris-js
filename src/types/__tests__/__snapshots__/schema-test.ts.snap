// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Type System: Schema Validity A Schema must contain uniquely named types rejects a Schema which defines an object type twice 1`] = `"There can be only one type named \\"SameName\\"."`;
